ciso_interior_forcing shouldn't set DIC_ & DOC_loc

mnlevy1981 opened the pull request:
The DIC_loc_fields and DOC_loc_fields components of
marbl_interior_share_type are initialized to c0 below level kmt in
marbl_setup_local_tracers() [since all values in tracer_local = c0 for
k>kmt]. By not setting these two components to c0 in marbl_ciso_mod, we
can change the intent of the interior_share argument from intent(inout)
to intent(in).

mnlevy1981 committed:
ciso_interior_forcing shouldn't set DIC_ & DOC_loc

The DIC_loc_fields and DOC_loc_fields components of
marbl_interior_share_type are initialized to c0 below level kmt in
marbl_setup_local_tracers() [since all values in tracer_local = c0 for
k>kmt]. By not setting these two components to c0 in marbl_ciso_mod, we
can change the intent of the interior_share argument from intent(inout)
to intent(in).

mnlevy1981 commented:
Note the change to `marbl_mod.F90` (stripping a comment): we actually need `marbl_interior_share` to be `intent(inout)` in `marbl_set_interior_forcing()` because it gets passed to `marbl_export_interior_shared_variables()` as an `intent(inout)` variable in line 1216; the comment was incorrect.

klindsay28 commented:
Could `marbl_interior_share`, and the other _share vars for that matter, be local variables in `marbl_set_interior_forcing` of dimension `domain%km`? I'm mentioning this because I don't see these vars being used outside of `marbl_set_interior_forcing` and its callees. I guess we might not want to do that if

1. there is a performance hit from the allocation going on (happening automatically under the covers),
2. we envision needing it outside of `marbl_set_interior_forcing`

mnlevy1981 commented:
I think that would be a good change to make -- in my mind, components of the `marbl_interface_class` either need to be available to the GCM or are part of the interface for thread safety (that is probably not a complete list, but I doubt any categories I may be missing apply to `interior_share`). Moving `interior_share` to a local variable would keep thread safety, and the GCM certainly doesn't need to access this derived type.

I had originally thought that we should wait until after we resolve #86 because our split of `marbl_mod.F90` (and, I imagine, `marbl_ciso_mod.F90`) could potential result in calling `marbl_ciso_set_interior_forcing()` from the `marbl_interface:set_interior_forcing` instead of from `marbl_set_interior_forcing()`... but then we would just declare the local variable in the interface routine instead... so I'll implement this suggestion and push the update if there is not a performance hit.

mnlevy1981 committed:
Remove interior_share from interface_class

We don't need interior_share to be part of the interface_class, it can
be a local variable in marbl_mod:marbl_set_interior_forcing()

mnlevy1981 committed:
Remove more share_types for interface_class

zooplankton_share, autotroph_share, and surface_forcing_share can all be
local variables in marbl_mod:marbl_set_interior_forcing(). Note that
particulate_share still needs to belong to the interface class because
it may be set during initialization (if ladjust_bury_coeff = .false.)
and the accessed during set_interior_forcing as well.

mnlevy1981 committed:
Put surface_forcing_share back in interface_class

Due to the way memory is allocated in the surface_forcing_share_type
(via a constructor call), surface_forcing_share needs to be in the
interface class.

mnlevy1981 committed:
fix intent() for marbl_particulate_share

in marbl_ciso_mod:marbl_ciso_set_interior_forcing,
marbl_particulate_share should be intent(in) not intent(inout). This
matches the other share_type variables in the subroutine.

mnlevy1981 commented:
It looks like `particulate_share` and `surface_forcing_share` need to stay in the interface because of how they are initialized, but `interior_share`, `zooplankton_share`, and `autotroph_share` do not benefit from being part of the interface so they have been moved to local variables in `marbl_set_interior_forcing`.

This may change if we move the `domain%km` dimension inside the type -- we have talked about doing that for some datatypes but I don't remember off the top of my head if these share types are included in that design. For now, though, I think taking these variables out of the interface is a good decision.

klindsay28 reviewed (approved):
looks good
